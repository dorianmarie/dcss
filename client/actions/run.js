import {
  GET_RUN_SUCCESS,
  GET_RUN_ERROR,
  GET_RUNS_SUCCESS,
  GET_RUNS_ERROR,
  GET_RUN_DATA_SUCCESS,
  GET_RUN_DATA_ERROR,
  SAVE_RUN_EVENT_SUCCESS,
  SAVE_RUN_EVENT_ERROR,
  SET_RUN,
  SET_RUN_SUCCESS,
  SET_RUN_ERROR
} from './types';

export let getRun = (scenario_id, cohort_id, chat_id) => async dispatch => {
  try {
    const base = `/api/runs/new-or-existing/scenario/${scenario_id}`;
    const cohort = cohort_id ? `/cohort/${cohort_id}` : '';
    const chat = chat_id ? `/chat/${chat_id}` : '';
    const url = `${base}${cohort}${chat}`;
    const res = await (await fetch(url)).json();

    if (res.error) {
      throw res;
    }
    const { run } = res;
    dispatch({ type: GET_RUN_SUCCESS, run });
    return run;
  } catch (error) {
    dispatch({ type: GET_RUN_ERROR, error });
    return null;
  }
};

export let getRunByIdentifiers = (
  scenario_id,
  cohort_id,
  chat_id
) => async dispatch => {
  try {
    const url = `/api/runs/by-identifiers/scenario/${scenario_id}/cohort/${cohort_id}/chat/${chat_id}`;
    const res = await (await fetch(url)).json();

    if (res.error) {
      throw res;
    }
    const { run } = res;
    dispatch({ type: GET_RUN_SUCCESS, run });
    return run;
  } catch (error) {
    dispatch({ type: GET_RUN_ERROR, error });
    return null;
  }
};

export let getRuns = () => async dispatch => {
  try {
    const res = await (await fetch('/api/runs')).json();

    if (res.error) {
      throw res;
    }
    const { runs } = res;
    dispatch({ type: GET_RUNS_SUCCESS, runs });
    return runs;
  } catch (error) {
    dispatch({ type: GET_RUNS_ERROR, error });
    return null;
  }
};

export let setRun = (id, data) => async dispatch => {
  dispatch({ type: SET_RUN, run: { id, ...data } });
  try {
    const body = JSON.stringify(data);
    const res = await (await fetch(`/api/runs/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json'
      },
      body
    })).json();

    if (res.error) {
      throw res;
    }

    const { run } = res;

    dispatch({ type: SET_RUN_SUCCESS, run });
    return run;
  } catch (error) {
    dispatch({ type: SET_RUN_ERROR, error });
    return null;
  }
};

export let getRunData = run_id => async dispatch => {
  try {
    const endpoint = `/api/runs/${run_id}`;
    const res = await (await fetch(endpoint)).json();

    if (res.error) {
      throw res;
    }

    const { prompts, responses } = res;

    dispatch({ type: GET_RUN_DATA_SUCCESS, prompts, responses });
    return { prompts, responses };
  } catch (error) {
    dispatch({ type: GET_RUN_DATA_ERROR, error });
    return null;
  }
};

export let saveRunEvent = (run_id, name, data) => async dispatch => {
  try {
    const timestamp = Date.now();
    const url = location.href;
    const context = {
      timestamp,
      url,
      ...data
    };

    const body = JSON.stringify({
      name,
      context
    });

    const res = await (await fetch(`/api/runs/${run_id}/event/${name}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body
    })).json();

    if (res.error) {
      throw res;
    }

    const { event } = res;
    dispatch({ type: SAVE_RUN_EVENT_SUCCESS, name, event });
    return event;
  } catch (error) {
    dispatch({ type: SAVE_RUN_EVENT_ERROR, error });
    return null;
  }
};
