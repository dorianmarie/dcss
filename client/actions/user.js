import Crypto from 'crypto-js';
import {
  GET_USER_SUCCESS,
  GET_USER_ERROR,
  SET_USER_SUCCESS,
  SET_USER_ERROR
} from './types';

export const signUp = params => async dispatch => {
  try {
    let { email = '', username = '', password = '' } = params;
    let data = {};

    if (email) {
      data.email = email.trim();
    }

    if (username) {
      data.username = username.trim();
    }

    if (password) {
      /* SESSION_SECRET is "embedded" by webpack */
      data.password = Crypto.AES.encrypt(password, SESSION_SECRET).toString();
    }

    const body = JSON.stringify(data);

    const res = await (await fetch('/api/auth/', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body
    })).json();

    if (res.error) {
      throw res;
    }
    const { user } = res;

    dispatch({ type: GET_USER_SUCCESS, user });
    return user;
  } catch (error) {
    dispatch({ type: GET_USER_ERROR, error });
    return null;
  }
};

export let getUser = () => async dispatch => {
  try {
    const res = await (await fetch('/api/auth/')).json();

    if (res.error) {
      throw res;
    }
    const { user } = res;

    dispatch({ type: GET_USER_SUCCESS, user });
    return user;
  } catch (error) {
    dispatch({ type: GET_USER_ERROR, error });
    return null;
  }
};

export let setUser = params => async dispatch => {
  try {
    if (Object.values(params).length) {
      if (params.password) {
        /* SESSION_SECRET is "embedded" by webpack */
        params.password = Crypto.AES.encrypt(
          params.password,
          SESSION_SECRET
        ).toString();
      }
      const body = JSON.stringify(params);
      const res = await (await fetch('/api/auth/', {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json'
        },
        body
      })).json();

      if (res.error) {
        throw res;
      }
      const { user } = res;

      dispatch({ type: SET_USER_SUCCESS, user });
      return user;
    }
    return null;
  } catch (error) {
    dispatch({ type: SET_USER_ERROR, error });
    return null;
  }
};

export let resetPassword = params => async () => {
  try {
    if (Object.values(params).length) {
      params.email = Crypto.AES.encrypt(
        params.email,
        SESSION_SECRET
      ).toString();
      params.origin = location.origin;
      const body = JSON.stringify(params);
      const res = await (await fetch('/api/auth/reset', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body
      })).json();

      if (res.error) {
        throw res;
      }
      const { reset } = res;
      return reset;
    }
    return null;
  } catch (error) {
    return null;
  }
};
