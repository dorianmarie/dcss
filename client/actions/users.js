import {
  GET_USERS_BY_PERMISSION_ERROR,
  GET_USERS_BY_PERMISSION_SUCCESS,
  GET_USERS_COUNT_ERROR,
  GET_USERS_COUNT_SUCCESS,
  GET_USERS_ERROR,
  GET_USERS_SUCCESS
} from './types';

export let setUsers = users => async dispatch => {
  dispatch({
    type: GET_USERS_SUCCESS,
    users
  });
};

export let getUsers = (limit = 'all') => async dispatch => {
  /*
    limit = "all" | "available"
  */
  try {
    const res = await (await fetch(`/api/roles/${limit}`)).json();

    if (res.error) {
      throw res;
    }

    const { users = [] } = res;
    dispatch({ type: GET_USERS_SUCCESS, users });
    return users;
  } catch (error) {
    dispatch({ type: GET_USERS_ERROR, error });
    return null;
  }
};

export let getUsersCount = (limit = 'all') => async dispatch => {
  /*
    limit = "all" | "available"
  */
  try {
    const res = await (await fetch(`/api/roles/${limit}/count`)).json();

    if (res.error) {
      throw res;
    }
    const count = Number(res.count);

    dispatch({ type: GET_USERS_COUNT_SUCCESS, count });
    return count;
  } catch (error) {
    dispatch({ type: GET_USERS_COUNT_ERROR, error });
    return null;
  }
};

/**
 * getUsersByPermission dispatch is not used by any reducer, because the returned values do not
 * need to be stored in memory.
 * @param  {string} permission The permission to filter users on.
 * @return {array}             An array of users that have this permission.
 */
export let getUsersByPermission = permission => async dispatch => {
  try {
    const res = await (await fetch(
      `/api/roles/user/permission/${permission}`
    )).json();
    if (res.error) {
      throw res;
    }
    const { users } = res;

    void GET_USERS_BY_PERMISSION_SUCCESS;

    return users;
  } catch (error) {
    dispatch({ type: GET_USERS_BY_PERMISSION_ERROR, error });
    return null;
  }
};
