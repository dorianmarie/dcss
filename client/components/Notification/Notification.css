.notification {
  z-index: 9999;
  position: fixed;
  top: 1em;
  right: 1em;
}

.notification__visible {
  pointer-events: auto;
  opacity: 1;
}

.notification__invisible {
  pointer-events: none;
  opacity: 0;
}

.n__container {
  min-width: 22vw;
  max-width: 33vw;
}

.n__container .content .header {
  margin-bottom: 0.25em !important;
}

.n__container .content .header > i.icon {
  font-size: 1em !important;
}

.n__container .content {
  overflow: hidden !important;
  text-overflow: ellipsis !important;
}

@media only screen and (max-width: 767px) {
  .n__container {
    min-width: unset;
    max-width: 100vw;
  }
}
