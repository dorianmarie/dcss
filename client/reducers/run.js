import {
  GET_RUN_SUCCESS,
  GET_RUNS_SUCCESS,
  SET_RUN_SUCCESS,
  SET_RUN
} from '@actions/types';

import { runInitialState } from './initial-states';

export const run = (state = runInitialState, action) => {
  const { run, type } = action;

  switch (type) {
    case GET_RUN_SUCCESS:
    case SET_RUN_SUCCESS:
    case SET_RUN:
      return {
        ...state,
        ...run
      };
    default:
      return state;
  }
};

export const runs = (state = [], action) => {
  const { runs, type } = action;

  switch (type) {
    case GET_RUNS_SUCCESS: {
      const seen = {
        /*
        run.id: index
         */
      };
      return [...state, ...runs]
        .reduce((accum, run) => {
          // If we've already seen this run, then it was already
          // in "state" and we should replace it with the newer version.
          const index = seen[run.id];
          if (index !== undefined) {
            accum[index] = run;
          } else {
            seen[run.id] = accum.length;
            accum.push(run);
          }
          return accum;
        }, [])
        .sort((a, b) => a.id < b.id);
    }
    default:
      return state;
  }
};

export const runsById = (state = {}, action) => {
  const { run, runs, type } = action;

  switch (type) {
    case GET_RUN_SUCCESS:
    case SET_RUN_SUCCESS:
    case GET_RUNS_SUCCESS: {
      if (typeof runs !== 'undefined') {
        const runsById = runs.reduce((accum, run) => {
          accum[run.id || run.run_id] = run;
          return accum;
        }, {});

        return {
          ...state,
          ...runsById
        };
      }

      if (typeof run !== 'undefined') {
        return {
          ...state,
          [run.id || run.run_id]: run
        };
      }
      return state;
    }
    default:
      return state;
  }
};
