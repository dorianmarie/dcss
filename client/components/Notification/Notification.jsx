import { ToastContainer as Notification } from '@components/Toast';
import './Notification.css';

export default Notification;
