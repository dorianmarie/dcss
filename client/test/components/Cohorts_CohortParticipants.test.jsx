import React from 'react';
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useLayoutEffect: jest.requireActual('react').useEffect
}));

import {
  fetchImplementation,
  mounter,
  reduxer,
  serialize,
  snapshotter,
  state
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';

import {
  act,
  fireEvent,
  prettyDOM,
  render,
  screen,
  waitFor
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

/** @GENERATED: BEGIN **/

import {
  GET_COHORT_SUCCESS,
  SET_COHORT_SUCCESS,
  GET_RUNS_SUCCESS,
  GET_USER_SUCCESS,
  GET_USERS_SUCCESS
} from '../../actions/types';
import * as cohortActions from '../../actions/cohort';
import * as runActions from '../../actions/run';
import * as userActions from '../../actions/user';
import * as usersActions from '../../actions/users';
jest.mock('../../actions/cohort');
jest.mock('../../actions/run');
jest.mock('../../actions/user');
jest.mock('../../actions/users');

import Layout from '@utils/Layout';
jest.mock('@utils/Layout', () => {
  return {
    ...jest.requireActual('@utils/Layout'),
    isForMobile: jest.fn(() => false),
    isNotForMobile: jest.fn(() => true)
  };
});

import copy from 'copy-text-to-clipboard';
jest.mock('copy-text-to-clipboard', () => {
  return jest.fn();
});

import Storage from '@utils/Storage';
jest.mock('@utils/Storage', () => {
  return {
    ...jest.requireActual('@utils/Storage'),
    get: jest.fn(),
    set: jest.fn()
  };
});

import { notify } from '@components/Notification';
jest.mock('@components/Notification', () => {
  return {
    ...jest.requireActual('@components/Notification'),
    notify: jest.fn()
  };
});

import CohortParticipants from '../../components/Cohorts/CohortParticipants.jsx';
/** @GENERATED: END **/

const original = JSON.parse(JSON.stringify(state));
let container = null;
let commonProps = null;
let commonState = null;

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

afterAll(() => {
  jest.restoreAllMocks();
});

beforeEach(() => {
  jest.useFakeTimers();
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);

  /** @GENERATED: BEGIN **/

  Layout.isForMobile = jest.fn();

  Layout.isForMobile.mockImplementation(() => false);

  Layout.isNotForMobile = jest.fn();

  Layout.isNotForMobile.mockImplementation(() => true);

  cohortActions.getCohort = jest.fn();

  cohortActions.getCohort.mockImplementation(() => async dispatch => {
    const cohort = {
      id: 1,
      created_at: '2020-08-31T14:01:08.656Z',
      name: 'A New Cohort That Exists Within Inline Props',
      is_archived: false,
      runs: [
        {
          id: 11,
          user_id: 333,
          scenario_id: 99,
          created_at: '2020-03-28T19:44:03.069Z',
          updated_at: '2020-03-31T17:01:43.139Z',
          ended_at: '2020-03-31T17:01:43.128Z',
          consent_id: 8,
          consent_acknowledged_by_user: true,
          consent_granted_by_user: true,
          referrer_params: null,
          cohort_id: 1,
          run_id: 11
        }
      ],
      scenarios: [99],
      users: [
        {
          username: 'super',
          personalname: 'Super User',
          email: 'super@email.com',
          id: 999,
          roles: ['participant', 'super_admin'],
          is_anonymous: false,
          is_super: true,
          progress: {
            completed: [1],
            latestByScenarioId: {
              1: {
                is_complete: true,
                event_id: 1909,
                created_at: 1602454306144,
                generic: 'arrived at a slide.',
                name: 'slide-arrival',
                url: 'http://localhost:3000/cohort/1/run/99/slide/1'
              }
            }
          }
        },
        {
          username: 'facilitator',
          personalname: 'Facilitator User',
          email: 'facilitator@email.com',
          id: 555,
          roles: ['participant', 'facilitator', 'researcher', 'owner'],
          is_anonymous: false,
          is_super: false,
          is_owner: true,
          progress: {
            completed: [],
            latestByScenarioId: {
              1: {
                is_complete: false,
                scenario_id: 99,
                event_id: 1905,
                created_at: 1602454306144,
                generic: 'arrived at a slide.',
                name: 'slide-arrival',
                url: 'http://localhost:3000/cohort/1/run/99/slide/1'
              }
            }
          }
        },
        {
          username: 'researcher',
          personalname: 'Researcher User',
          email: 'researcher@email.com',
          id: 444,
          roles: ['participant', 'researcher'],
          is_anonymous: false,
          is_super: false,
          progress: {
            completed: [],
            latestByScenarioId: {
              1: {
                is_complete: false,
                scenario_id: 99,
                event_id: 1904,
                created_at: 1602454306144,
                generic: 'arrived at a slide.',
                name: 'slide-arrival',
                url: 'http://localhost:3000/cohort/1/run/99/slide/1'
              }
            }
          }
        },
        {
          username: 'participant',
          personalname: 'Participant User',
          email: 'participant@email.com',
          id: 333,
          roles: ['participant'],
          is_anonymous: false,
          is_super: false,
          progress: {
            completed: [],
            latestByScenarioId: {
              1: {
                is_complete: false,
                scenario_id: 99,
                event_id: 1903,
                created_at: 1602454306144,
                generic: 'arrived at a slide.',
                name: 'slide-arrival',
                url: 'http://localhost:3000/cohort/1/run/99/slide/1'
              }
            }
          }
        },
        {
          username: 'anonymous',
          personalname: '',
          email: '',
          id: 222,
          roles: ['participant'],
          is_anonymous: true,
          is_super: false,
          progress: {
            completed: [],
            latestByScenarioId: {
              1: {
                is_complete: false,
                scenario_id: 99,
                event_id: 1902,
                created_at: 1602454306144,
                generic: 'arrived at a slide.',
                name: 'slide-arrival',
                url: 'http://localhost:3000/cohort/1/run/99/slide/1'
              }
            }
          }
        }
      ],
      roles: ['super', 'facilitator'],
      usersById: {
        999: {
          username: 'super',
          personalname: 'Super User',
          email: 'super@email.com',
          id: 999,
          roles: ['participant', 'super_admin'],
          is_anonymous: false,
          is_super: true,
          progress: {
            completed: [1],
            latestByScenarioId: {
              1: {
                is_complete: true,
                event_id: 1909,
                created_at: 1602454306144,
                generic: 'arrived at a slide.',
                name: 'slide-arrival',
                url: 'http://localhost:3000/cohort/1/run/99/slide/1'
              }
            }
          }
        },
        555: {
          username: 'facilitator',
          personalname: 'Facilitator User',
          email: 'facilitator@email.com',
          id: 555,
          roles: ['participant', 'facilitator', 'researcher', 'owner'],
          is_anonymous: false,
          is_super: false,
          is_owner: true,
          progress: {
            completed: [],
            latestByScenarioId: {
              1: {
                is_complete: false,
                scenario_id: 99,
                event_id: 1905,
                created_at: 1602454306144,
                generic: 'arrived at a slide.',
                name: 'slide-arrival',
                url: 'http://localhost:3000/cohort/1/run/99/slide/1'
              }
            }
          }
        },
        444: {
          username: 'researcher',
          personalname: 'Researcher User',
          email: 'researcher@email.com',
          id: 444,
          roles: ['participant', 'researcher'],
          is_anonymous: false,
          is_super: false,
          progress: {
            completed: [],
            latestByScenarioId: {
              1: {
                is_complete: false,
                scenario_id: 99,
                event_id: 1904,
                created_at: 1602454306144,
                generic: 'arrived at a slide.',
                name: 'slide-arrival',
                url: 'http://localhost:3000/cohort/1/run/99/slide/1'
              }
            }
          }
        },
        333: {
          username: 'participant',
          personalname: 'Participant User',
          email: 'participant@email.com',
          id: 333,
          roles: ['participant'],
          is_anonymous: false,
          is_super: false,
          progress: {
            completed: [],
            latestByScenarioId: {
              1: {
                is_complete: false,
                scenario_id: 99,
                event_id: 1903,
                created_at: 1602454306144,
                generic: 'arrived at a slide.',
                name: 'slide-arrival',
                url: 'http://localhost:3000/cohort/1/run/99/slide/1'
              }
            }
          }
        },
        222: {
          username: 'anonymous',
          personalname: '',
          email: '',
          id: 222,
          roles: ['participant'],
          is_anonymous: true,
          is_super: false,
          progress: {
            completed: [],
            latestByScenarioId: {
              1: {
                is_complete: false,
                scenario_id: 99,
                event_id: 1902,
                created_at: 1602454306144,
                generic: 'arrived at a slide.',
                name: 'slide-arrival',
                url: 'http://localhost:3000/cohort/1/run/99/slide/1'
              }
            }
          }
        }
      }
    };
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    return cohort;
  });
  usersActions.getUser = jest.fn();

  usersActions.getUser.mockImplementation(() => async dispatch => {
    const user = {
      username: 'facilitator',
      personalname: 'Facilitator User',
      email: 'facilitator@email.com',
      id: 555,
      roles: ['participant', 'facilitator', 'researcher', 'owner'],
      is_anonymous: false,
      is_super: false,
      is_owner: true,
      progress: {
        completed: [],
        latestByScenarioId: {
          1: {
            is_complete: false,
            scenario_id: 99,
            event_id: 1905,
            created_at: 1602454306144,
            generic: 'arrived at a slide.',
            name: 'slide-arrival',
            url: 'http://localhost:3000/cohort/1/run/99/slide/1'
          }
        }
      }
    };
    dispatch({ type: GET_USER_SUCCESS, user });
    return user;
  });
  usersActions.getUsers = jest.fn();

  usersActions.getUsers.mockImplementation(() => async dispatch => {
    const users = [
      {
        username: 'super',
        personalname: 'Super User',
        email: 'super@email.com',
        id: 999,
        roles: ['participant', 'super_admin'],
        is_anonymous: false,
        is_super: true,
        progress: {
          completed: [1],
          latestByScenarioId: {
            1: {
              is_complete: true,
              event_id: 1909,
              created_at: 1602454306144,
              generic: 'arrived at a slide.',
              name: 'slide-arrival',
              url: 'http://localhost:3000/cohort/1/run/99/slide/1'
            }
          }
        }
      },
      {
        username: 'facilitator',
        personalname: 'Facilitator User',
        email: 'facilitator@email.com',
        id: 555,
        roles: ['participant', 'facilitator', 'researcher', 'owner'],
        is_anonymous: false,
        is_super: false,
        is_owner: true,
        progress: {
          completed: [],
          latestByScenarioId: {
            1: {
              is_complete: false,
              scenario_id: 99,
              event_id: 1905,
              created_at: 1602454306144,
              generic: 'arrived at a slide.',
              name: 'slide-arrival',
              url: 'http://localhost:3000/cohort/1/run/99/slide/1'
            }
          }
        }
      },
      {
        username: 'researcher',
        personalname: 'Researcher User',
        email: 'researcher@email.com',
        id: 444,
        roles: ['participant', 'researcher'],
        is_anonymous: false,
        is_super: false,
        progress: {
          completed: [],
          latestByScenarioId: {
            1: {
              is_complete: false,
              scenario_id: 99,
              event_id: 1904,
              created_at: 1602454306144,
              generic: 'arrived at a slide.',
              name: 'slide-arrival',
              url: 'http://localhost:3000/cohort/1/run/99/slide/1'
            }
          }
        }
      },
      {
        username: 'participant',
        personalname: 'Participant User',
        email: 'participant@email.com',
        id: 333,
        roles: ['participant'],
        is_anonymous: false,
        is_super: false,
        progress: {
          completed: [],
          latestByScenarioId: {
            1: {
              is_complete: false,
              scenario_id: 99,
              event_id: 1903,
              created_at: 1602454306144,
              generic: 'arrived at a slide.',
              name: 'slide-arrival',
              url: 'http://localhost:3000/cohort/1/run/99/slide/1'
            }
          }
        }
      },
      {
        username: 'anonymous',
        personalname: '',
        email: '',
        id: 222,
        roles: ['participant'],
        is_anonymous: true,
        is_super: false,
        progress: {
          completed: [],
          latestByScenarioId: {
            1: {
              is_complete: false,
              scenario_id: 99,
              event_id: 1902,
              created_at: 1602454306144,
              generic: 'arrived at a slide.',
              name: 'slide-arrival',
              url: 'http://localhost:3000/cohort/1/run/99/slide/1'
            }
          }
        }
      }
    ];
    dispatch({ type: GET_USERS_SUCCESS, users });
    return users;
  });

  Storage.get.mockImplementation(key => {
    return null;
  });

  /** @GENERATED: END **/

  commonProps = {};
  commonState = JSON.parse(JSON.stringify(original));
});

afterEach(() => {
  jest.runOnlyPendingTimers();
  jest.useRealTimers();
  jest.resetAllMocks();
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  commonProps = null;
  commonState = null;
});

test('CohortParticipants', () => {
  expect(CohortParticipants).toBeDefined();
});

test('Render 1 1', async done => {
  /** @GENERATED: BEGIN **/
  const Component = CohortParticipants;
  const props = {
    ...commonProps,
    id: 1,
    authority: { isFacilitator: true, isParticipant: true }
  };

  const state = {
    ...commonState
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);
  /** @GENERATED: END **/

  const { asFragment } = render(<ConnectedRoutedComponent {...props} />);
  expect(asFragment()).toMatchSnapshot();
  await screen.findByRole('navigation', { name: /pagination navigation/i });
  expect(asFragment()).toMatchSnapshot();

  done();
});

/* INJECTION STARTS HERE */

test('Search participants', async done => {
  const Component = CohortParticipants;

  const props = {
    ...commonProps,
    id: 1,
    authority: { isFacilitator: true, isParticipant: true }
  };

  const state = {
    ...commonState
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  await render(<ConnectedRoutedComponent {...props} />);
  expect(serialize()).toMatchSnapshot();
  await screen.findByRole('navigation', { name: /pagination navigation/i });
  expect(serialize()).toMatchSnapshot();

  jest.advanceTimersByTime(1);
  expect(setTimeout).toHaveBeenCalledTimes(1);
  expect(serialize()).toMatchSnapshot();

  const searchInput = await screen.findByLabelText('Search participants');

  // "researcher"
  userEvent.type(searchInput, 'resea{enter}');
  expect(serialize()).toMatchSnapshot();

  jest.advanceTimersByTime(1);
  expect(setTimeout).toHaveBeenCalledTimes(2);
  expect(serialize()).toMatchSnapshot();

  userEvent.type(searchInput, '{selectall}{backspace}');
  expect(serialize()).toMatchSnapshot();

  // "owner"
  userEvent.type(searchInput, 'owner{enter}');
  expect(serialize()).toMatchSnapshot();

  jest.advanceTimersByTime(1);
  expect(setTimeout).toHaveBeenCalledTimes(3);
  expect(serialize()).toMatchSnapshot();

  // "researcher@email.com"
  userEvent.type(searchInput, '{selectall}{backspace}');
  userEvent.type(searchInput, 'researcher@email.com{enter}');
  expect(serialize()).toMatchSnapshot();

  jest.advanceTimersByTime(1);
  expect(setTimeout).toHaveBeenCalledTimes(4);
  expect(serialize()).toMatchSnapshot();

  userEvent.type(
    searchInput,
    '{selectall}{backspace}nothing will match this{enter}'
  );
  expect(serialize()).toMatchSnapshot();

  userEvent.type(searchInput, '{selectall}{backspace}');
  expect(serialize()).toMatchSnapshot();

  jest.advanceTimersByTime(1);
  expect(setTimeout).toHaveBeenCalledTimes(5);
  expect(serialize()).toMatchSnapshot();

  done();
});

test('Pagination', async done => {
  const Component = CohortParticipants;

  const props = {
    ...commonProps,
    rowsPerPage: 2,
    id: 1,
    authority: { isFacilitator: true, isParticipant: true }
  };

  const state = {
    ...commonState
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  await render(<ConnectedRoutedComponent {...props} />);
  expect(serialize()).toMatchSnapshot();
  const pagination = await screen.findByRole('navigation', {
    name: /pagination navigation/i
  });
  expect(serialize()).toMatchSnapshot();

  userEvent.click(await screen.getByText('2'));
  expect(serialize()).toMatchSnapshot();

  userEvent.click(await screen.getByText('3'));
  expect(serialize()).toMatchSnapshot();

  userEvent.click(await screen.getByText('1'));
  expect(serialize()).toMatchSnapshot();

  done();
});

test('Modal close', async done => {
  const Component = CohortParticipants;

  const onClose = jest.fn();
  const props = {
    ...commonProps,
    onClose,
    rowsPerPage: 2,
    id: 1,
    authority: { isFacilitator: true, isParticipant: true }
  };

  const state = {
    ...commonState
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  await render(<ConnectedRoutedComponent {...props} />);
  expect(serialize()).toMatchSnapshot();
  await screen.findByRole('navigation', { name: /pagination navigation/i });
  expect(serialize()).toMatchSnapshot();

  userEvent.click(await screen.findByLabelText('Close'));

  expect(onClose).toHaveBeenCalledTimes(1);
  done();
});

test('Super admin', async done => {
  const Component = CohortParticipants;

  Storage.get.mockImplementation(key => {
    return {
      refresh: true
    };
  });

  const props = {
    ...commonProps,
    id: 1,
    authority: { isFacilitator: true, isParticipant: true }
  };

  const state = {
    ...commonState,
    user: {
      username: 'super',
      personalname: 'Super User',
      email: 'super@email.com',
      id: 999,
      roles: ['participant', 'super_admin'],
      is_anonymous: false,
      is_super: true
    }
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  await render(<ConnectedRoutedComponent {...props} />);
  expect(serialize()).toMatchSnapshot();
  await screen.findByRole('navigation', { name: /pagination navigation/i });
  expect(serialize()).toMatchSnapshot();

  done();
});

test('Participant only', async done => {
  const Component = CohortParticipants;

  const props = {
    ...commonProps,
    id: 1,
    authority: { isFacilitator: false, isParticipant: true }
  };

  const state = {
    ...commonState,
    user: {
      username: 'participant',
      personalname: 'Participant User',
      email: 'participant@email.com',
      id: 333,
      roles: ['participant'],
      is_anonymous: false,
      is_super: false
    }
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  await render(<ConnectedRoutedComponent {...props} />);
  expect(serialize()).toMatchSnapshot();

  done();
});

test('Researcher only', async done => {
  const Component = CohortParticipants;

  const props = {
    ...commonProps,
    id: 1,
    authority: { isFacilitator: false, isParticipant: true }
  };

  const state = {
    ...commonState,
    user: {
      username: 'researcher',
      personalname: 'Researcher User',
      email: 'researcher@email.com',
      id: 444,
      roles: ['participant', 'researcher'],
      is_anonymous: false,
      is_super: false
    }
  };
  const ConnectedRoutedComponent = reduxer(Component, props, state);

  await render(<ConnectedRoutedComponent {...props} />);
  expect(serialize()).toMatchSnapshot();

  done();
});
