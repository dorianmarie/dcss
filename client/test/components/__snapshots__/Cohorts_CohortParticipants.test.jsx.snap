// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal close 1`] = `<DocumentFragment />`;

exports[`Modal close 2`] = `<DocumentFragment />`;

exports[`Pagination 1`] = `<DocumentFragment />`;

exports[`Pagination 2`] = `<DocumentFragment />`;

exports[`Pagination 3`] = `<DocumentFragment />`;

exports[`Pagination 4`] = `<DocumentFragment />`;

exports[`Pagination 5`] = `<DocumentFragment />`;

exports[`Participant only 1`] = `<DocumentFragment />`;

exports[`Render 1 1 1`] = `<DocumentFragment />`;

exports[`Render 1 1 2`] = `<DocumentFragment />`;

exports[`Researcher only 1`] = `<DocumentFragment />`;

exports[`Search participants 1`] = `<DocumentFragment />`;

exports[`Search participants 2`] = `<DocumentFragment />`;

exports[`Search participants 3`] = `<DocumentFragment />`;

exports[`Search participants 4`] = `<DocumentFragment />`;

exports[`Search participants 5`] = `<DocumentFragment />`;

exports[`Search participants 6`] = `<DocumentFragment />`;

exports[`Search participants 7`] = `<DocumentFragment />`;

exports[`Search participants 8`] = `<DocumentFragment />`;

exports[`Search participants 9`] = `<DocumentFragment />`;

exports[`Search participants 10`] = `<DocumentFragment />`;

exports[`Search participants 11`] = `<DocumentFragment />`;

exports[`Search participants 12`] = `<DocumentFragment />`;

exports[`Search participants 13`] = `<DocumentFragment />`;

exports[`Super admin 1`] = `<DocumentFragment />`;

exports[`Super admin 2`] = `<DocumentFragment />`;
